### ISSUE TITLE
How to ONLY check types and emit unmodified JavaScript without the types?

### ISSUE THREAD

**Issue by jsejcksn** [microsoft/TypeScript#30140]
I want to write code in a `.ts` file which is simply JavaScript having TypeScript types, and instruct the compiler to emit a JavaScript file containing my code without the types, but not transpiled and otherwise unmodified in any way—basically a setting like `onlyCheckTypesAndDoNothingElse`. **How to do this?**

There is [a page about JSDoc support](https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript). This is kind of what I want, except I actually want to use TypeScript types.

Like this
``` ts
function example(a: number, b: number): number {
    return a + b;
}

```

instead of this
``` js
/**
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
function example(a, b) {
    return a + b;
}
```

I read through the following [docs](https://www.typescriptlang.org/docs/home.html) pages without finding anything related to my question:

  - [TypeScript in 5 minutes
](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html)
  - [Migrating from JavaScript](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html)
  - [tsconfig.json
](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) (By the way, the wiki [FAQ](https://github.com/Microsoft/TypeScript/wiki) link to [tsconfig.json](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/tsconfig.json.md) is broken.)
  - [Compiler Options](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

It appears there might be a combination of compiler options to achieve what I want, but the idea of a deep dive to understand every compiler option and the interactions between them seems like a massive effort to get started.

**Comment by nattthebear** [comment]
Assuming your settings are otherwise default, target esnext should be sufficient, except for the weird occasional thing like decorators.

**Comment by jsejcksn** [comment]
@nattthebear Thanks for responding. Your comment illustrates exactly what I mean. I want my code absolutely unmodified except for removal of TypeScript types.

**Comment by dragomirtitian** [comment]
@jsejcksn I think the babel typescript plugin just removes types annotations. If you just use that (without any other plugins) it should leave the code unmodified. But it will not perform type checking, you could use ` --noEmit` to get the compiler to just to type checking.

**Comment by RyanCavanaugh** [comment]
> I want my code absolutely unmodified except for removal of TypeScript types

`target: esnext` is the closest thing available.

"Remove only types" isn't a thing because we have some features (decorators, enums, namespaces) which can't simply be *removed*. But if you don't use those features, then `target: esnext` will print your code with identical semantics.

If you need **literally just exactly the type annotations removed** (why? - curious) then I think you'd have to use Babel or possibly even some other tool.

**Comment by jsejcksn** [comment]
@dragomirtitian Thank you for offering an idea for a solution. I would not like to add yet another dependency and build step, but I will investigate further.

@RyanCavanaugh Thanks for responding. Your suggestion still transforms even the most basic code, replacing `const` with `var`.

```shell
$ ls -a
.               ..              .DS_Store       tsconfig.json
$ cat tsconfig.json 
{
  "compilerOptions": {
    "target": "esnext"
  }
}
$ echo "let foo: string = 'bar';" > example.ts
$ tsc example.ts 
$ cat example.js 
var foo = 'bar';
```

> why? - curious

I want to have the benefits of static type-checking, but I don't want my code transformed—as simple as that.

If I have non-type errors in my code, I want them emitted as-is. I understand that TypeScript is a powerful tool that does increasingly more and more, but isn't the core purpose to add static typing to JavaScript? I don't understand why it can't do just that. 😞 I want to use TypeScript, but I don't want the compiler transforming my code.

**Comment by weswigham** [comment]
> Your suggestion still transforms even the most basic code, replacing `const` with `var`.

...whut. That's not right at all. Or it shouldn't happen that way, anyway. Any target > es6 should be preserving `const`/`let`. Oh, your invocation doesn;t actually use your tsconfig, that's why. A tsconfig isn't like a `.babelrc`, it doesn't implicitly apply to any invocations in that folder. If you invoke `tsc` with no arguments we look for one, but if you explicitly pass a file, we do not. Run `tsc example.ts --target esnext` or just run `tsc`.

**Comment by jsejcksn** [comment]
@weswigham When I re-ran using just `tsc` it didn't transform. 🙌 Thanks for the correction.

My understanding is that what you guys are telling me is:

  - **it's not currently possible to do what I want with TypeScript**, _but_
  - depending on which JavaScript features I use or don't use, the emitted output will be essentially what I want

This leaves me with two questions:

  1. If I use `tsc --target esnext`, what kinds of transformations do I need to watch out for?
  2. How do I create a suggestion for the product roadmap to include a compiler option for "no code transformations except removal of type annotations"?

**Comment by weswigham** [comment]
> what kinds of transformations do I need to watch out for?

Decorators and class fields will always still be transpiled, since those are both part of the `ts` transform (plus enums and namespaces, but those aren't JS of any kind, so I don't think that should surprise you).

AFAIK they're considered part of TS because they didn't stablize (and still haven't) in an at-all timely manner and the way TS interprets them is different (majorly for decorators, minorly for fields) than current proposals.

**Comment by RyanCavanaugh** [comment]
> 2\. How do I create a suggestion for the product roadmap to include a compiler option for "no code transformations except removal of type annotations"?

I think what you're asking for is a mode that disables things which aren't JS or TypeScript (e.g. enums) ? Something akin to this was suggested in #2261 but I think we could revive a flag for "Don't let me do things which have code emit".

More details about exactly which aspects of this are important would be helpful in pointing you toward an ideal solution. Like, if you write `1_000_000` (legal numeric separators in ES Next) and target ES6, is emitting `1000000` OK, or is that "transforming" and needs to be banned? What's the motivation?

**Comment by jsejcksn** [comment]
@RyanCavanaugh 

> ...if you write `1_000_000` (legal numeric separators in ES Next) and target ES6, is emitting `1000000` OK, or is that "transforming" and needs to be banned? What's the motivation?

No, this would not be ok. I don't want any kind of transformations.

> I think what you're asking for is a mode that disables things which aren't JS or TypeScript (e.g. enums) ?

No, not exactly. Imagine wanting to write plain JavaScript for an environment in which the parser is set to bleeding edge (ESNext and—ideally—anything the JS engine can handle, even with a flag: @weswigham [Public class fields on Chrome platform status](https://www.chromestatus.com/feature/6001727933251584)), but also wanting static typing. That's what I'm asking for. I just want to write JavaScript with static types, have the code checked for type errors only, and then have the type annotations removed on emission.

If I want to transpile down to a specific target environment, I recognize that there are already options for that in TypeScript and Babel, but I want an option that does not do this.

What I'm describing would essentially disable everything else and just enable "type-only checking". I would tell the parser: "Hey, I am giving you vanilla JS + type annotations. Give me back the vanilla JS (with feedback about how bad I am at programming when there are type errors that I didn't catch when writing it)."

I don't know how else to explain given my current knowledge about TypeScript.

**Comment by RyanCavanaugh** [comment]
I really do want to hear about your underlying motivations for this mode. What you're describing is so close to the existing `esnext` target + simply avoiding TS-specific constructs (of which there are very few) that it seems like there must be some other requirement you're trying to fulfill. I don't want to go down the road of speccing out some special flag only to find out that e.g. you need all comments *before* type annotations *but not after* to be removed, etc.

**Comment by jsejcksn** [comment]
@RyanCavanaugh I don't think I know enough about TS-specific constructs to answer your question.

My motivation is to start writing my JavaScript in `.ts` files immediately and progressively use more features as I learn about them, resting assured that the code emitted during compile will include **all of my code exactly as I wrote it** (except the non-JS, TS-specific annotations that I add along the way).

**Comment by jsejcksn** [comment]
@RyanCavanaugh Do you have any feedback at this point? I am wondering about the status of this issue (e.g. "in internal discussion", "won't fix / not in scope—use Babel", etc.).

**Comment by jsejcksn** [comment]
Why was this closed without a response?

**Comment by jsejcksn** [comment]
@RyanCavanaugh Still waiting on a response

**Comment by RyanCavanaugh** [comment]
@jsejcksn you posted a question and we answered it; we're not able to give extensive 1:1 discussions much further than that. If you want to start a new suggestion that outlines exactly what kind of new functionality you'd like to see, please do!

**Comment by randomekek** [comment]
For what it's worth, somebody forked typescript and made it accept types as inline comments. This means it's possible to not even compile the file and directly use it with the browser. You can also type check on the side.

https://github.com/luciotato/plus-typescript/tree/plus
